### DISCUSSION TITLE
App architecture question

### DISCUSSION THREAD

**Discussion by Marakai** [derekhe/msfs2020-google-map#145] [Q&A] [answered]
Hi Derek,

I just had a simple question regarding the overall architecture of the application. Am I correct in thinking that:

- FS2020 sends its geo requests, normally to Bing Maps via https/443, but instead to
- the fake entries in the modified `hosts` file, which direct to `127.0.0.1`
- on `127.0.0.1` we have NGINX listening on 443
- NGINX proxies the queries, does whatever URL rewriting required (haven't looked at the config in detail yet) and then sends queries to
- Flask, which has a few RESTful API functions, is listening on 39871 and it is then the one that
- sends queries to Google Maps properly translated
- the Google Maps data is translated into something that looks like it came from Bing
- then it all goes back up the pipeline

Is that right?

**Comment by derekhe** [comment, marked as answer]
Yes, correct. 
The nginx is used to handle the requests and do url remapping. I found a problem that Flask can't handle http2.0 requests, so the photogrammetry data will not load. Use nginx can avoid this issue. But if any server can do the proxy well, we can get rid of nginx.

**Reply by Marakai** [reply]
Yes, the fewer components you need to bundle, the easier it becomes to install for users. Interesting that Flask can't handle 2.0... is this known and do they have a roadmap to fix it?

**Reply by Marakai** [reply]
Also, I'm sure you've seen https://stackoverflow.com/questions/38878880/serving-python-flask-rest-api-over-http2#38891504 ? Especially the answer about Twisted and especially Quart - which is Flask for http2? EDIT: correction. Looking at the docs, Quart is mostly a rewrite of Flask using asyncio, however it specifically states that it supports HTTP/2.
